src/scheduler.rs

```diff
@@ -0,0 +1,1013 @@
+    PodIdentityNotParseable { pod_id: String },
+}
+
+/// Returns a Vec of pod identities according to the replica per role+group pair from `eligible_nodes`.
+/// # Arguments
+/// * `app_name` : Application name
+/// * `instance` : Service instance
+/// * `eligible_nodes` : Eligible nodes grouped by role and groups.
+pub fn generate_ids(
+    app_name: &str,
+    instance: &str,
+    eligible_nodes: &EligibleNodesForRoleAndGroup,
+) -> Vec<PodIdentity> {
+    let mut id: u16 = 1;
+    let mut generated_ids = vec![];
+    for (role_name, groups) in eligible_nodes {
+        for (group_name, eligible_nodes) in groups {
+            let ids_per_group = eligible_nodes
+                .replicas
+                .map(usize::from)
+                .unwrap_or_else(|| eligible_nodes.nodes.len());
+            for _ in 1..ids_per_group + 1 {
+                generated_ids.push(PodIdentity {
+                    app: app_name.to_string(),
+                    instance: instance.to_string(),
+                    role: role_name.clone(),
+                    group: group_name.clone(),
+                    id: id.to_string(),
+                });
+                id += 1;
+            }
+        }
+    }
+
+    generated_ids
+}
+
+#[derive(
+    Clone, Debug, Default, Deserialize, Eq, Hash, JsonSchema, Ord, PartialEq, PartialOrd, Serialize,
+)]
+#[serde(rename_all = "camelCase")]
+#[serde(try_from = "String")]
+#[serde(into = "String")]
+pub struct PodIdentity {
+    pub app: String,
+    pub instance: String,
+    pub role: String,
+    pub group: String,
+    pub id: String,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NodeIdentity {
+    pub name: String,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PodToNodeMapping {
+    mapping: BTreeMap<PodIdentity, NodeIdentity>,
+}
+
+pub trait PodPlacementHistory {
+    fn find(&self, pod_id: &PodIdentity) -> Option<&NodeIdentity>;
+    fn find_all(&self, pods: &[&PodIdentity]) -> Vec<Option<&NodeIdentity>> {
+        pods.iter().map(|p| self.find(*p)).collect()
+    }
+
+    fn update(&mut self, pod_id: &PodIdentity, node_id: &NodeIdentity);
+}
+
+pub struct K8SUnboundedHistory<'a> {
+    pub client: &'a Client,
+    pub history: PodToNodeMapping,
+    modified: bool,
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct RoleGroupEligibleNodes {
+    node_set: BTreeMap<String, BTreeMap<String, Vec<NodeIdentity>>>,
+}
+
+/// Represents the successful result of a `schedule()`
+/// It contains the current scheduled pods and the remaining pods to be scheduled (`new_mapping`)
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct SchedulerState {
+    current_mapping: PodToNodeMapping,
+    new_mapping: PodToNodeMapping,
```

Comment:
I actually like the name in the comment (new_mapping -> remaining_mapping or unscheduled_mapping)?
I think new mapping is a bit confusion, since the mapping isnt new at all?